# SilenceREIN

In this work, we constructed a regulatory element interaction network (REIN) by integrating chromatin structure data and regulatory elements data. We extracted the REIN topology features using a graph neural network model. Various linear chromatin signatures are also integrated with a deep neural network model. We name our model as SilenceREIN.

## 1. Platform and Dependency

### 1.1 Platform

+ ubuntu18.04
+ RTX 3080(10GB)

### 1.2 Dependency

| Requirements      | Release   |
|:------------------|:----------|
| CUDA              | 11.3      |
| Python            | 3.9.0     |
| torch             | 1.12.1    |
| torch_geometric   | 2.2.0     |
| pybedtools        | 0.9.0     |
| pyBigWig          | 0.3.18    |
| numpy             | 1.23.4    |
| scikit-learn      | 1.1.3     |
| tqdm              | 4.64.1    |
| pandas            | 1.5.1     |
| matplotlib        | 3.6.2     |


We need to download the bedtools tool for building the dataset, for example, for Ubuntu systems, the command to download bedtools is:
> apt-get install bedtools
>
For more information please see [Installation â€” bedtools 2.31.0 documentation](https://bedtools.readthedocs.io/en/latest/content/installation.html)


## 2. Project Catalog Structure

### 2.1 src
> This folder stores the code files. 

+ myclass.py
  > This file contains some classes used when the dataset is built and loaded.
+ build_REIN.py 
  > This file generates Regulatory Element Interaction Network and constructs it as an instance of the REIN class. The generated files will be stored in the `data/REIN` folder.
+ build_dataset.py
  > This file generates the dataset used in training. The generated files will be stored in the `dataset` folder.
  > 
  > The dataset comprises topological information and one-dimensional genomic information. The topological information is stored in the `raw` folder within the specified directory, while the one-dimensional and genomic information is saved as `x-labeled.txt.gz` or `x-unlabeled.txt.gz`. The `raw-txt` folder contains a set of description files for each file in the `raw` folder.
+ model.py
  > This file contains the code for the SilenceREIN model.
+ cross_validation.py
  > This file containing the code for running cross validation for SilenceREIN. During the running of the code, log files are generated and stored in the `result/log` folder.
+ train.py
  > This file containing the code for training a SilenceREIN model.
+ predict.py
  > This file contains code for predicting silencer from unlabeled nodes.
+ performance.py
  > This file contains the code for evaluating model performance in cross-validation.
+ utils.py
  > This file contains the code for reading and writing files, and loading training data.
+ apa_data.py
  > The file generate contacts for APA.

### 2.2 data
| Folder name | Descriptions                                                                                                                    |
|:------------|:--------------------------------------------------------------------------------------------------------------------------------|
| ChIA-PET    | The chromatin loops generated from ChIA-PET.                                                                                    |
| HiChIP      | The chromatin loops generated from HiChIP.                                                                                      |
| ChIP-seq    | The histone ChIP-seq data, and TF ChIP-seq data.                                                                                |
| CREs        | Files in bed format for silencer, non-silencer, enhancer, promoter. Transformation of the genome version into hg38 by liftover. |
| hg38        | Human genome.                                                                                                                   |
| REIN        | REIN generated by `build_REIN.py` with the name `graph.pkl`.                                                                    |
| tmp         | The intermediate files generated by build_REIN.py, i.e. the DNA sequences of the anchor points and CREs.                        |
| APA         | The contacts between promoters and predicted silencers.                                                                         |
| Annotation  | The annotations of ChromHMM, Segway and the fully automated annotation model.                                                   |


### 2.3 dataset
> The dataset used in training and cross-validation. 
> The topological information is stored in the `raw` folder within the specified directory, while the one-dimensional and genomic information of 
> labeled nodes and unlabeled nodes is saved as `x-labeled.txt.gz` and `x-unlabeled.txt.gz`, respectively. The `raw-txt` folder contains a set of description files for each file in the `raw` folder.
> `K562-ChIA-PET-REIN.txt` provides a description of the dataset for K562 cells constructed based on ChIA-PET data. It includes node IDs, genomic coordinates, node types, training masks, and more.

### 2.4 result
+ logs
  > The logs generated in training and cross-validation, which record confusion matrices, performance metrics, etc.
+ models
  > This folder contains the trained models and the models obtained from each fold of the cross validations.
+ record
  > This folder contains the predicted and true values for test samples in the cross-validation generated by `cross_validation.py`.
+ predict
  > This folder contains the predicted silencer generated by `predict.py`.

## 3. Workflow

### 3.1 Build REIN
#### 3.1.1 Download hg38 from UCSC

> cd data/hg38
>
> wget https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/latest/hg38.fa.gz
> gunzip hg38.fa.gz

#### 3.1.2.Download histone ChIP-seq data and Transcription factor ChIP-seq data

> cd data/K562/ChIP-seq/Set1/histone-ChIP-seq
> 
> xargs -L 1 curl -O -J -L < files.txt

> cd data/K562/ChIP-seq/Set2/TF-ChIP-seq
> 
> xargs -L 1 curl -O -J -L < files.txt

To download ChIP-seq data from Set2, you can move to the Set2 directory and execute the same command.

This will take some time to complete the download.

#### 3.1.3 build REIN

> python src/build_REIN.py

##### Optional parameters
* --cell: Default=`K562`. Specify the cell line.
* --HiC: Default=`ChIA-PET`. Specify the chromatin conformation data type used.

This will take some time and after this step you will get the REIN, which is stored in the folder `data/REIN`.


### 3.2 Build dataset

#### 3.2.1 Generate Dataset

> python src/build_dataset.py

##### Optional parameters
* --cell: Default=`K562`. Specify the cell line.
* --HiC: Default=`ChIA-PET`. Specify the chromatin conformation data type used.

This will generate the dataset in this work.


### 3.3 Cross Validation, Training, and Prediction

#### 3.3.1 Hyper-parameters description
| Hyper-parameter | Type   | Default | Description                                                           |
|:----------------|:-------|:--------|:----------------------------------------------------------------------|
| learning rate   | float  | 0.0001  | Batch Size, the default value is recommended.                         |
| weight decay    | float  | 0.0001  | Weight decay, the default value is recommended.                       |                                                                       |
| class weight    | float  | 1.2     | Class weight, the default value is recommended.                       |                                                                       ||
| gamma           | float  | 0.95    | As the loss rises, the learning rate will be multiplied by the gamma. |

#### 3.3.2 Cross validation
> python src/cross_validation.py

This will save models and logs in `result/models` and `result/logs`, respectively. The confusion matrix and performance will be reported each epoch.

##### Optional parameters
* --model_name: Default=`SilenceREIN-CV`. The name of model in cross-validation.
* --dataset: Default=`Set1`. The dataset used for cross validation. The other option is `Set2`.
* --epoch: Default=40. The number of training epoch.
* --batch_size: Default=128. The batch size of this training.
* --lr: Default=0.0001. The initial learning rate.
* --weight_decay: Default=1e-4. The weight decay for this training.
* --weight: Default=1.2. The class weight.
* --seed: Default=0. Random seed for this cross validation.
* --ChIP_seq: Default=`All`. The ChIP-seq data. This parameter can be selected for All, H3K27ac, etc.

#### 3.3.3 Training
> python src/train.py

This will save models and logs in `result/models` and `result/logs`, respectively.

##### Optional parameters
* --model_name: Default=`SilenceREIN`. The name of model.
* --epoch: Default=40. The number of training epoch.
* --batch_size: Default=128. The batch size of this training.
* --lr: Default=0.0001. The initial learning rate.
* --weight_decay: Default=1e-4. The weight decay for this training.
* --weight: Default=1.2. The class weight.


#### 3.3.4 Prediction
> python src/predict.py

This will predict unlabeled nodes from REIN, and the predicted silencers will be saved.

##### Optional parameters
* --model_name: Default=`SilenceREIN`. The name of model. Which was used in our study.


### Figures in this study
#### Figure3:
Plot the performance of SilenceREIN against [CNN](https://github.com/ncbi/SilencerEnhancerPredict), [DeepSilencer](https://github.com/xy-chen16/DeepSilencer), and gkmSVM. We fed the data used in this study to CNN, DeepSilencer, and gkmSVM for multiple rounds of 5-fold cross-validation for plotting in Figure 3.

For gkmSVM, we use the following command:
> gkmsvm_kernel('pos.fasta','neg.fasta', 'kernel.out')
> 
> gkmsvm_trainCV('kernel.out','pos.fasta','neg.fasta',svmfnprfx='svm', outputCVpredfn='cvpred.out', outputROCfn='roc.out')

#### Figure4:
Plot the performance of SilenceREIN vs. SilenceREIN-alt.

#### Figure5:
Plot the performance of the model when only topological and linear information is used.

#### Figure6:
UMAP visualization of the hidden layer (before MLP) features of SilenceREIN, SilenceREIN-alt, SilenceREIN (No Network features), and SilenceREIN-alt (No Network features).


#### Figure7:
Plot the performance of SilenceREIN against ChromHMM, Segway and the fully automated annotation model. 
To achieve this, begin by running `get_performance.py` to get the prediction of the annotation models.

#### Figure8:
Visualize the distribution of predicted silencers across ChromHMM, Segway, and the fully automated annotation model's annotations.
To achieve this, begin by running `get_percent.py` to calculate the distribution of predicted silencers within each annotation.

#### FigureS9:
Present the distribution of predicted silencers on ChromHMM and Segway annotations. These annotations are based on our primary benchmarking dataset and have been interpreted by a pre-trained fully automated annotation model.

#### Figure10:
Calculate motif significance scores and plot statistical histograms. We set the P-value less than 1e-6. Motifs data downloaded from JASPAR.
The motifs scan results are from [FIMO](https://meme-suite.org/meme/tools/fimo).

#### FigureS1:
Plot the performance of the model when only one type of ChIP-seq data is used.

#### FigureS2:
Create performance comparison plots for SilenceREIN (and SilenceREIN-alt) against CNN, DeepSilencer, and gkmSVM based on ChIA-PET data in the HepG2 cell line.

#### FigureS3:
Create performance comparison plots for SilenceREIN (and SilenceREIN-alt) against CNN, DeepSilencer, and gkmSVM based on HiChIP data in the K562 cell line.
